torch/utils/data/datapipes/_typing.py

```diff
@@ -422,25 +465,43 @@ def wrap_generator(*args, **kwargs):
                 raise
 
         namespace['__iter__'] = wrap_generator
-    else:
+    else:  # ``__iter__`` of IterDataPipe is NOT a generator function
         # IterDataPipe is an iterator with both ``__iter__`` and ``__next__``
-        if '__next__' in namespace:
+        # And ``__iter__`` may or may not return `self`
+        if '__next__' in namespace:  # If `__next__` exists, put a wrapper around it
```

Comment:
Is there way that we can check if the `__iter__` is returning `self`?
And, it seems we never set `datapipe._valid_iterator_id` to 1 for this case (`_set_datapipe_valid_iterator_id` is never invoked)

Reply:
> Is there way that we can check if the `__iter__` is returning `self`?

I can have another look. A bit tricky since we are within the hook, we might have to call it and examine the output.

> And, it seems we never set `datapipe._valid_iterator_id` to 1 for this case (`_set_datapipe_valid_iterator_id` is never invoked)

Yes, it is only set later by `wrap_iter` if `iter` is called. In theory the user can have two active references (i.e. `dp` and `it = iter(dp)`. This is true unless we ban them from using `iter` completely.

Reply:
> A bit tricky since we are within the hook, we might have to call it and examine the output.

If that's the case, we are probably not going to do so

> Yes, it is only set later by `wrap_iter` if `iter` is called. In theory the user can have two active references (i.e. `dp` and `it = iter(dp)`. This is true unless we ban them from using `iter` completely.

What does that mean? Can you provide an example of expected behavior for such scenario?

Reply:
This test case right here:
https://github.com/pytorch/pytorch/blob/87bba0f26e4463fb766bf736ddc09757b1053370/test/test_datapipe.py#L2188-L2214

Notice that both `source_dp` and `it1` are valid here:
```python
self.assertEqual(0, next(it1))
self.assertEqual(1, next(source_dp))
```

Reply:
Yea, I think when `__iter__` returns `self`, any attempt to update/restrict ID doesn't work too well in notebook mode.

I originally proposed that if `__iter__` returns `self`, we do not need to track any ID, and should just let the users call `__iter__` as many time as they would like. That has a different downside too, since users might accidentally have multiple references to it without realizing.

Reply:
I believe we are proposing a `reset` function, right? Then, whenever `__iter__` is invoked, the `reset` will handle re-initialization for all variables like `buffer`. As long as users have proper `reset` function, it should be fine.

Reply:
The only issue is when they re-run a cell 2 without re-running cell 1 in the example below. An exception will be raised the second time they run it because `dp` will become invalidated. The workaround is to override `reset` such that `_valid_iterator_id` is set back to `None` during reset.

Cell 1:
```python
dp = CustomDP()  # Its `__iter__` returns `self`
```
Cell 2:
```python
it = iter(dp)
```

Reply:
Wait, I remember we discussed an alternative to hook __iter__ method and to check if it returns self. And in such a scenario wrap the return into a new IterableContainer of some sort to track validation. Let me know if you need sudocode for an explanation.

Reply:
It does get wrapped around by `IteratorDecorator` which delegates method calls to the iterator object (if `self` is returned, then the DataPipe is the iterator object).

Please look at this comment: https://github.com/pytorch/pytorch/pull/70479#discussion_r874217619